04.ls/ls.rb

```diff
@@ -1,57 +1,154 @@
 # frozen_string_literal: true
 
 require 'optparse'
+require 'etc'
 
 COLUMN = 3
-SPACE = 2
+BLOCK_SIZE = 4
+
+FILE_TYPE = {
+  '01' => 'p',
+  '02' => 'c',
+  '04' => 'd',
+  '10' => '-',
+  '12' => 'l',
+  '14' => 's'
+}.freeze
+
+PERMISSION = {
+  '0' => '---',
+  '1' => '--x',
+  '2' => '-w-',
+  '3' => '-wx',
+  '4' => 'r--',
+  '5' => 'r-x',
+  '6' => 'rw-',
+  '7' => 'rwx'
+}.freeze
+
+IS_AUTHORITY = {
+  '0' => false,
+  '1' => true,
+  '2' => true,
+  '4' => true
+}.freeze
+
+INDEX_AUTHORITY = {
+  '1' => 8,
+  '2' => 5,
+  '4' => 2
+}.freeze
+
+AUTHORITY_TYPE = {
+  '1' => { 'x' => 't', '-' => 'T' },
+  '2' => { 'x' => 's', '-' => 'S' },
+  '4' => { 'x' => 's', '-' => 'S' }
+}.freeze
 
 def main
-  parsed_options = parse_options
-  target_dir = ARGV.empty? ? '.' : ARGV[0]
-  entry_names = Dir.entries(target_dir)
-  sorted_names = sort_names(entry_names, parsed_options)
-  filtered_names = filter_names(sorted_names, parsed_options)
-  puts output(filtered_names)
+  options = parse_options
+  target_dir_path = ARGV.empty? ? '.' : ARGV[0]
+  names = Dir.entries(target_dir_path)
+  sorted_names = sort_names(names, options)
+  loaded_names = load_names(sorted_names, target_dir_path, options)
+  formatted_names = options[:l] ? format_attributes(loaded_names) : format_names(loaded_names)
+  puts formatted_names
 end
 
 def parse_options
   options = {}
   OptionParser.new do |opts|
     opts.on('-a') { options[:a] = true }
     opts.on('-r') { options[:r] = true }
+    opts.on('-l') { options[:l] = true }
   end.parse!(ARGV)
   options
 end
 
 def sort_names(names, options)
-  options[:r] ? names.sort.reverse : names.sort
+  sorted_names = names.sort_by(&:upcase)
+  options[:r] ? sorted_names.reverse : sorted_names
 end
 
-def filter_names(names, options)
-  options[:a] ? names : names.reject { |name| name.start_with?('.') }
+def load_names(names, path, options)
+  loaded_names = options[:a] ? names : names.reject { |name| name.start_with?('.') }
+  options[:l] ? load_attributes(loaded_names, path) : loaded_names
 end
 
-def output(names)
+def load_attributes(names, path)
+  total_block_size = 0
+
+  loaded_attributes = names.each.map do |name|
+    name_path = File.absolute_path(name, path)
+    file_stat = File::Stat.new(name_path)
+    total_block_size += (file_stat.size.to_f / file_stat.blksize).ceil * BLOCK_SIZE
+    attributes = []
+
+    attributes << load_type_and_permission(file_stat)
+    attributes << file_stat.nlink
+    attributes << Etc.getpwuid(file_stat.uid).name
+    attributes << Etc.getgrgid(file_stat.gid).name
+    attributes << file_stat.size
+    attributes << file_stat.mtime.strftime('%b %e %H:%M')
+    attributes << name
```

Comment:
```suggestion
    attributes << name
    attributes
```

`<<`の戻り値をブロックの戻り値にするのはちょっと意図が明確ではないので、配列そのものを明示的に最後に返す方がいいと思います。

そもそもの話として、配列リテラルを返す、という方法もあります。

```ruby
[
  load_type_and_permission(file_stat),
  # ...
  name
]
```